### PR TITLE
click relay equipment and firearm patches.

### PR DESCRIPTION
**Pull request by Ghommie**
:cl:
fix: Fixes alt / shift / ctrl clicks (e.g. examinating something) not properly working with click relay equipment such as turrets.
fix: The burst-fire toogle verb for the M56D should now only show to the operator.
tweak: firearm verbs will now also check for the inactive hand, as opposed of simply failing if it's not present in the active one.
/:cl:

The M56D one is quite goofy I admit, I haven't documented myself well enough on the possibility of action buttons and Byond is fairly limited on verbs. Shoul be more than a band aid for #519.
Tested locally, but may still be prone to changes. #383 already contains the updated version of the tank's `handle_click()`.

### CODE REVIEW

File: code/modules/cm_marines/smartgun_mount.dm [hunk 1]
```diff
@@ -244,16 +244,17 @@
...
-			operator.unset_interaction()
-		SetLuminosity(0)
-		STOP_PROCESSING(SSobj, src)
-		. = ..()
+/obj/machinery/m56d_hmg/Destroy() //Make sure we pick up our trash.
+	if(operator)
+		operator.unset_interaction()
+	SetLuminosity(0)
+	STOP_PROCESSING(SSobj, src)
+	. = ..()
```
Comment (DominikPanic) [inline code/modules/cm_marines/smartgun_mount.dm]:
```suggestion
	return ..()
```

File: code/_onclick/human.dm [hunk 2]
```diff
@@ -10,8 +10,8 @@
 	var/last_chew = 0
 
 /mob/living/carbon/human/click(var/atom/A, var/list/mods)
-	if(interactee)
-		return interactee.handle_click(src, A, mods)
+	if(interactee?.handle_click(src, A, mods))
+		return
```
Comment (spookydonut) [inline code/_onclick/human.dm]: so no longer returning a value?

File: code/modules/cm_marines/equipment/sentries.dm [hunk 3]
```diff
@@ -1061,39 +1061,32 @@
 
 //Direct replacement to new proc. Everything works.
 /obj/machinery/marine_turret/handle_click(mob/living/carbon/human/user, atom/A, params)
-	if(!operator || !istype(user) || operator != user)
-		return FALSE
-	if(istype(A, /obj/screen))
-		return FALSE
-	if(!manual_override)
+	var/list/modifiers = params2list(params) //Only single clicks.
+	if(modifiers["middle"] || modifiers["shift"] || modifiers["alt"] || modifiers["ctrl"])
```
Comment (spookydonut) [inline code/modules/cm_marines/equipment/sentries.dm]: this is really ugly, shouldnt see this outside of code/_onclick/

File: code/modules/cm_marines/smartgun_mount.dm [hunk 4]
```diff
@@ -441,32 +442,23 @@
 
 // New proc for MGs and stuff replaced handle_manual_fire(). Same arguements though, so alls good.
 /obj/machinery/m56d_hmg/handle_click(mob/living/carbon/human/user, atom/A, var/list/mods)
-	if(!operator)
-		return FALSE
-	if(operator != user)
-		return FALSE
-	if(istype(A,/obj/screen))
+	if(mods["middle"] || mods["shift"] || mods["alt"] || istype(A,/obj/screen) || !operator || operator != user)
```
Comment (spookydonut) [inline code/modules/cm_marines/smartgun_mount.dm]: same here

File: code/modules/projectiles/updated_projectiles/gun_helpers.dm [hunk 5]
```diff
@@ -743,25 +730,26 @@ should be alright.
...
-	var/obj/item/weapon/gun/W = usr.get_active_held_item()
-
-	if(!istype(W))
+	var/obj/item/weapon/gun/G = get_active_firearm(usr)
+	if(!G)
 		return
+	src = G
 
-	W.rail?.activate_attachment(W, usr)
+	rail?.activate_attachment(src, usr)
```
Comment (spookydonut) [inline code/modules/projectiles/updated_projectiles/gun_helpers.dm]:
```suggestion
	rail?.activate_attachment(G, usr)
```
no need to assign this to src

File: code/modules/cm_marines/equipment/sentries.dm (hunk 3 above)
Comment (Ghommie) [reply code/modules/cm_marines/equipment/sentries.dm]: If it weren't for some special mod click interaction on some click relay equipment, I'd have done that at the root already.

File: code/_onclick/human.dm (hunk 2 above)
Comment (Ghommie) [reply code/_onclick/human.dm]: Well, better play safe and make it return a value then.

**Comment by Ghommie** [comment]
Ok, I have fixed the firearm pop up menu issues for real now

File: code/modules/projectiles/updated_projectiles/gun_helpers.dm [hunk 6]
```diff
@@ -738,9 +738,10 @@ should be alright.
 	var/obj/item/weapon/gun/G = get_active_firearm(usr)
 	if(!G)
 		return
-	src = G
 
-	G.rail?.activate_attachment(G, usr)
+	if(!G.rail)
+		to_chat(usr, "<span class='warning'>[src] does not have any usable rail attachment!</span>")
```
Comment (DominikPanic) [inline code/modules/projectiles/updated_projectiles/gun_helpers.dm]: Shouldn't you return here?

File: code/_onclick/human.dm [hunk 7]
```diff
@@ -10,8 +10,9 @@
 	var/last_chew = 0
 
 /mob/living/carbon/human/click(var/atom/A, var/list/mods)
-	if(interactee)
-		return interactee.handle_click(src, A, mods)
+	if(!istype(A,/obj/screen))
```
Comment (Rohesie) [inline code/_onclick/human.dm]: Do an early return here instead, please.

File: code/modules/cm_marines/smartgun_mount.dm [hunk 8]
```diff
@@ -244,16 +244,17 @@
...
 
-	Destroy() //Make sure we pick up our trash.
-		if(operator)
-			operator.unset_interaction()
-		SetLuminosity(0)
-		STOP_PROCESSING(SSobj, src)
-		. = ..()
+/obj/machinery/m56d_hmg/Destroy() //Make sure we pick up our trash.
+	if(operator)
+		operator.unset_interaction()
```
Comment (Rohesie) [inline code/modules/cm_marines/smartgun_mount.dm]:
```suggestion
		operator?.unset_interaction()
```
And remove the check.

File: code/modules/cm_marines/equipment/sentries.dm (hunk 3 above)
Comment (spookydonut) [reply code/modules/cm_marines/equipment/sentries.dm]: move the proc into a file in that directory

File: code/_onclick/handle_click.dm [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+		return FALSE
+	if(user.get_active_held_item())
+		to_chat(usr, "<span class='warning'>You need a free hand to shoot the [src].</span>")
+		return TRUE
+	target = A
+	if(!istype(target))
+		return FALSE
+	if(target.z != z || !target.z || !z || isnull(operator.loc) || isnull(loc))
+		return FALSE
+	if(get_dist(target,src.loc) > 15)
```
Comment (Rohesie) [inline code/_onclick/handle_click.dm]:
```suggestion
	if(get_dist(target, loc) > 15)
```

File: code/_onclick/handle_click.dm [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+	if(user.lying || !Adjacent(user) || user.is_mob_incapacitated())
+		user.unset_interaction()
+		return FALSE
+	if(user.get_active_held_item())
+		to_chat(usr, "<span class='warning'>You need a free hand to shoot the [src].</span>")
+		return TRUE
+	target = A
+	if(!istype(target))
+		return FALSE
+	if(target.z != z || !target.z || !z || isnull(operator.loc) || isnull(loc))
```
Comment (Rohesie) [inline code/_onclick/handle_click.dm]:
```suggestion
	if(isnull(loc) || isnull(operator.loc) || !z || !target.z || target.z != z)
```
Sanity checks first.

File: code/_onclick/handle_click.dm [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+	if(get_dist(target,src.loc) > 15)
+		return TRUE
+
+	if(mods["ctrl"])
+		burst_fire = !burst_fire
+		burst_fire_toggled = TRUE
+
+	var/angle = get_dir(src,target)
+	//we can only fire in a 90 degree cone
+	if((dir & angle) && target.loc != src.loc && target.loc != operator.loc)
```
Comment (Rohesie) [inline code/_onclick/handle_click.dm]:
```suggestion
	if((dir & angle) && target.loc != loc && target.loc != operator.loc)
```

File: code/_onclick/handle_click.dm [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+		process_shot()
+		return TRUE
+
+	return FALSE
+
+//No one but the gunner can gun
+//And other checks to make sure you aren't breaking the law
+/obj/vehicle/multitile/root/cm_armored/tank/handle_click(mob/living/user, atom/A, list/mods)
+
+	if(!can_use_hp(user)) return
```
Comment (Rohesie) [inline code/_onclick/handle_click.dm]: Same line control statement.

File: code/_onclick/handle_click.dm [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+	if(!hardpoints.Find(active_hp))
+		to_chat(user, "<span class='warning'>Please select an active hardpoint first.</span>")
+		return
+
+	var/obj/item/hardpoint/HP = hardpoints[active_hp]
+
+	if(!HP)
+		return
+
+	if(!HP.is_ready())
```
Comment (Rohesie) [inline code/_onclick/handle_click.dm]:
```suggestion
	if(!HP?.is_ready())
```
And remove the previous check.

**Pull request merged**